### PR TITLE
Ensure EC keys with a private key but without a public key can be created

### LINKED ISSUE #12612: OpenSSL 3.0.0-alpha6 6 Aug 2020 breaks perl Net::DNS::SEC ECDSA signature generation
The problem does not occur in the signing process itself.
This statement fails with an exception raised in checkerr():

  ```
	checkerr( EVP_DigestSignInit( ctx, NULL, md, NULL, pkey ) );

  ```
Returned value is 0, not 1 as expected.
Code works ok with alpha5 and all earlier releases.
I have not been following the daily snapshots, so unable to be more precise.

Problem occurs for both ECDSAP256SHA256 and ECDSAP384SHA384 signature generation.
Signature verification is not affected which should rule out problems with md.

### PR DESCRIPTION
**Pull request by mattcaswell**
In 1.1.1 and earlier it was possible to create EC_KEYs that did not have the public key in it. We need to ensure that this continues to work in 3.0.

Fixes #12612

### CODE REVIEW

File: providers/implementations/keymgmt/ec_kmgmt.c [hunk 1]
```diff
@@ -341,12 +341,8 @@ int common_import(void *keydata, int selection, const OSSL_PARAM params[],
      */
     if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) == 0)
         return 0;
-    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0
-            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) == 0)
-        return 0;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0)
-        ok = ok && ec_group_fromdata(ec, params);
+    ok = ok && ec_group_fromdata(ec, params);
```
Comment (levitte) [inline providers/implementations/keymgmt/ec_kmgmt.c:345]: Did you notice the `include_private` further down, passed to `ec_key_fromdata()`?  That has lost its importance (I'm not sure why it was even important to begin with), so I'd throw that away, including removing that last `ec_key_fromdata()` argument.

Comment (mattcaswell) [reply providers/implementations/keymgmt/ec_kmgmt.c:345]: Doesn't that mean that we would no longer pay any attention to the `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` selection parameter? i.e. we would only drive things off of whether the private key data was actually present in the params - regardless of what `selection` says. That doesn't seem quite right to me....and if we no longer consider that selection parameter important, then perhaps a bigger refactor is required?

Comment (levitte) [reply providers/implementations/keymgmt/ec_kmgmt.c:345]: Saying that `selection` is unimportant if we don't specifically check for `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` is a bit of an exaggeration, isn't it?  We do still check that a keypair is expected before calling `ec_key_fromdata()`, do we not?  However, I don't see the value with being that level of picky here if `ec_key_fromdata()` can discover on its own if there's a private key present in the data or not.

Comment (mattcaswell) [reply providers/implementations/keymgmt/ec_kmgmt.c:367]: Frankly, I'm quite reluctant to make the suggested change. I seriously dislike the inconsistency of paying attention to some of the section flags, but not others. To me that has a nasty smell about it. It also seems quite tangential to the point of this PR.

**Comment by mattcaswell** [comment]
Rebased to resolve a conflict with master. Please take a look!

**Comment by mattcaswell** [comment]
Fix up pushed addressing the no-deprecated build failure (due to the recent EC_KEY deprecation on master). I have no clue about the cmp test failure that occurred. It's difficult to see how that is related to these changes, so lets see if it happens again.

**Comment by t8m** [comment]
I think the cmp test failure is just intermittent thing due to the test parallelization - IMO it's some other test stepping on the cmp_test toes.

File: providers/implementations/keymgmt/ec_kmgmt.c (hunk 1 above)
Comment (levitte) [reply providers/implementations/keymgmt/ec_kmgmt.c:367]: [shrug]

**Comment by levitte** [comment]
> I think the cmp test failure is just intermittent thing due to the test parallelization - IMO it's some other test stepping on the cmp_test toes.

The competing cmp tests are probably `test_cmp_hdr` and `test_cmp_status`.  This diff should fix that:

``` diff
diff --git a/test/recipes/65-test_cmp_hdr.t b/test/recipes/65-test_cmp_hdr.t
index e013a6d820..9acb784d7e 100644
--- a/test/recipes/65-test_cmp_hdr.t
+++ b/test/recipes/65-test_cmp_hdr.t
@@ -12,7 +12,7 @@ use strict;
 use OpenSSL::Test qw/:DEFAULT data_file/;
 use OpenSSL::Test::Utils;
 
-setup("test_cmp_lib");
+setup("test_cmp_hdr");
 
 plan skip_all => "This test is not supported in a no-cmp build"
     if disabled("cmp");
diff --git a/test/recipes/65-test_cmp_status.t b/test/recipes/65-test_cmp_status.t
index 6a0a0c06b6..23f833f695 100644
--- a/test/recipes/65-test_cmp_status.t
+++ b/test/recipes/65-test_cmp_status.t
@@ -12,7 +12,7 @@ use strict;
 use OpenSSL::Test qw/:DEFAULT data_file/;
 use OpenSSL::Test::Utils;
 
-setup("test_cmp_lib");
+setup("test_cmp_status");
 
 plan skip_all => "This test is not supported in a no-cmp build"
     if disabled("cmp");
```

**Comment by mattcaswell** [comment]
Ping?

**Review by levitte — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**